core/lib/zksync_core/src/api_server/web3/pubsub.rs

```diff
@@ -238,118 +228,143 @@ impl EthSubscribe {
         self.events_sender = Some(sender);
     }
 
-    /// Assigns ID for the subscriber if the connection is open, returns error otherwise.
-    fn assign_id(
-        subscriber: typed::Subscriber<PubSubResult>,
-    ) -> Result<(typed::Sink<PubSubResult>, SubscriptionId), ()> {
-        let id = H128::random();
-        let sub_id = SubscriptionId::String(format!("0x{}", hex::encode(id.0)));
-        let sink = subscriber.assign_id(sub_id.clone())?;
-        Ok((sink, sub_id))
+    async fn reject(sink: PendingSubscriptionSink) {
+        sink.reject(ErrorObject::borrowed(
+            ErrorCode::InvalidParams.code(),
+            "Rejecting subscription - invalid parameters provided.",
+            None,
+        ))
+        .await;
     }
 
-    fn reject(subscriber: typed::Subscriber<PubSubResult>) {
-        subscriber
-            .reject(Error {
-                code: ErrorCode::InvalidParams,
-                message: "Rejecting subscription - invalid parameters provided.".into(),
-                data: None,
-            })
-            .unwrap();
+    async fn subscriber(
+        sink: SubscriptionSink,
+        subscription_type: SubscriptionType,
+        mut b: broadcast::Receiver<Vec<PubSubResult>>,
+        filter: Option<PubSubFilter>,
+    ) -> Result<(), anyhow::Error> {
+        let _guard = PUB_SUB_METRICS.active_subscribers[&subscription_type].inc_guard(1);
+        let closed = sink.closed().fuse();
+        tokio::pin!(closed);
+
+        loop {
+            tokio::select! {
+                new_items = b.recv() => {
+                    let new_items = new_items?;
+                    for item in new_items {
```

Comment:
Bikeshedding: IMO, it'd make sense to extract items handling to a separate function:

```rust
async fn handle_new_items(
    sink: &SubscriptionSink,
    subscription_type: SubscriptionType,
    new_items: Vec<PubSubResult>,
    filter: Option<&PubSubFilter>,
) -> Result<(), SendTimeoutError> {
    // ...
}
```

This would help readability and minimize the amount of code inside `tokio::select!` with potential formatting problems.

Reply:
done